ch5/ex-5-17-qsort-fields.c

```diff
@@ -0,0 +1,228 @@
+{
+	int i, rslt = 0;
+
+	for (i = 0; i < nsortflds && rslt == 0; i++)
+		rslt = (*compares[i])(sortdata1[i], sortdata2[i]);
+	return rslt;
+}
+
+void parseargs(int argc, char *argv[])
+{
+	int i;
+	int maxsortcount = argc - 1;
+
+	nsortflds = 0;
+	sortidxs = malloc(maxsortcount * sizeof(int));
+	compares = malloc(maxsortcount * sizeof(CHARCOMP));
+	folds = malloc(maxsortcount * sizeof(int));
+	dirsorts = malloc(maxsortcount * sizeof(int));
+
+	argv++;
+	for (i = 0; i < argc - 1; i++) {
+		char *arg = argv[i];
+
+		int fieldidx;
+		int numeric = 0;
+		int reverse = 0;
+		int (*compare)(char *, char *);
+		int fold = 0;
+		int dirsort = 0;
+
+		if (*arg == '-' && isdigit(*++arg)) {
+			fieldidx = atoi(arg) - 1;
+			while (*++arg) {
+				if (*arg == 'n')
+					numeric = 1;
+				else if (*arg == 'r')
+					reverse = 1;
+				else if (*arg == 'f')
+					fold = 1;
+				else if (*arg == 'd')
+					dirsort = 1;
+			}
+			if (numeric && reverse)
+				compare = numcmpr;
+			else if (numeric)
+				compare = numcmp;
+			else if (reverse)
+				compare = strcmpr;
+			else
+				compare = (CHARCOMP)strcmp;
+			maxsortidx =
+				fieldidx > maxsortidx ? fieldidx : maxsortidx;
+			sortidxs[nsortflds] = fieldidx;
+			compares[nsortflds] = compare;
+			folds[nsortflds] = fold;
+			dirsorts[nsortflds] = dirsort;
+			nsortflds++;
+		}
+	}
+}
+
+char **getsortdata(char *line)
+{
+	char *chr, *field;
+	int fieldidx, i;
+
+	char *fieldspace = malloc((strlen(line) + 1) * sizeof(char));
+	char **tmpfields = malloc((maxsortidx + 2) * sizeof(char *));
+	char **sortdata = malloc((nsortflds + 2) * sizeof(char *));
+
+	strcpy(fieldspace, line);
+
+	/* split line into fields and create temporary pointer for each field */
+	field = fieldspace;
+	fieldidx = 0;
+	for (chr = fieldspace; *chr; chr++) {
+		if (*chr == '\t' || *chr == '\n') {
+			*chr = '\0';
+			/*maxsortidx + 1 so last sort field is terminated */
+			if (fieldidx <= maxsortidx + 1)
+				tmpfields[fieldidx] = field;
+			field = chr + 1;
+			fieldidx++;
+		}
+	}
+
+	/* ptr to sort-by fields in priority order, apply -f and -d if needed */
+	for (i = 0; i < nsortflds; i++) {
+		field = tmpfields[sortidxs[i]];
+		if (folds[i])
+			fold(field);
+		if (dirsorts[i])
+			dirsort(field);
+		sortdata[i] = field;
+	}
+	sortdata[nsortflds] = line;
+	sortdata[nsortflds + 1] = fieldspace;
+
+	free(tmpfields);
+	return sortdata;
+}
+
+char *line(char *sortdata[])
+{
+	return sortdata[nsortflds];
+}
+
+char *fieldspace(char *sortdata[])
+{
+	return sortdata[nsortflds + 1];
+}
+
+void fold(char *value)
+{
+	while ((*value = toupper(*value)))
+		value++;
+}
+
+void dirsort(char *value)
+{
+	char c, *write;
+	write = value;
+	while ((c = *value++))
+		if (isalnum(c) || c == ' ' || c == '\n')
+			*write++ = c;
+	*write = '\0';
+}
+
+int strcmpr(char *s1, char *s2)
+{
+	return strcmp(s2, s1);
+}
+
+int numcmp(char *s1, char *s2)
+{
+	double v1 = atof(s1);
+	double v2 = atof(s2);
+
+	return v1 < v2 ? -1 : v1 > v2 ? 1 : 0;
+}
```

Comment:
This is fine.

But if you wanted to convert the floats outside the function, you could change it to 
```
int numcmp(void *s1, void *s2)
{
	double v1 = *(double*)s1;
	double v2 = *(double*)s2;
        ...
```
And the `strcmpr` version would be the same but with `char` instead of `double`.
That way the two functions would have the same signature and be assigned to the same function pointer.
Bypassing the type system when it gets in the way like this is called type erasure in C and is considered very cool.

Reply:
Definately going to have to think about/play with that one.